pkg/api/metrics.go

```diff
@@ -18,8 +122,95 @@ type Metrics struct {
 	ReceivedQueries     prometheus.Counter
 	FailedQueries       prometheus.Counter
 	QueryBatchDuration  prometheus.Histogram
-	CachedMetricNames   prometheus.CounterFunc
-	CachedLabels        prometheus.CounterFunc
 	InvalidReadReqs     prometheus.Counter
 	InvalidWriteReqs    prometheus.Counter
+	CachedMetricNames   prometheus.CounterFunc
+	HTTPRequestDuration *prometheus.HistogramVec
+	CachedLabels        prometheus.CounterFunc
+	MetricNamesCacheCap prometheus.GaugeFunc
+	LabelsCacheCap      prometheus.GaugeFunc
+}
+
+func InitMetrics() *Metrics {
+	if metrics != nil {
+		return metrics
+	}
+	prometheus.MustRegister(leaderGauge)
+	prometheus.MustRegister(receivedSamples)
+	prometheus.MustRegister(receivedQueries)
+	prometheus.MustRegister(sentSamples)
+	prometheus.MustRegister(failedSamples)
+	prometheus.MustRegister(failedQueries)
+	prometheus.MustRegister(invalidReadReqs)
+	prometheus.MustRegister(invalidWriteReqs)
+	prometheus.MustRegister(sentBatchDuration)
+	prometheus.MustRegister(queryBatchDuration)
+	prometheus.MustRegister(httpRequestDuration)
+	prometheus.MustRegister(ReadQueryLatency)
+	writeThroughput.Start()
+
+	metrics = &Metrics{
+		LeaderGauge:         leaderGauge,
+		ReceivedSamples:     receivedSamples,
+		FailedSamples:       failedSamples,
+		SentSamples:         sentSamples,
+		SentBatchDuration:   sentBatchDuration,
+		WriteThroughput:     writeThroughput,
+		LastRequestUnixNano: lastRequestUnixNano,
+		QueryBatchDuration:  queryBatchDuration,
+		FailedQueries:       failedQueries,
+		ReceivedQueries:     receivedQueries,
+		HTTPRequestDuration: httpRequestDuration,
+	}
+
+	return metrics
+}
+
+func InitClientMetrics(client *pgclient.Client) *Metrics {
```

Comment:
Optional additional refactor: I really think the code would look a lot better if this was in pkg/pgclient/metrics.go and looked like the existing pkg/pgmodel/metrics.go

There is no reason this needs to be part of of the metrics struct we pass around we can just make this local to pkg/pgclient

Reply:
Makes sense, will do it.